src/utils/deepClone.ts

```diff
@@ -0,0 +1,25 @@
+// ==LICENSE-BEGIN==
+// Copyright 2017 European Digital Reading Lab. All rights reserved.
+// Licensed to the Readium Foundation under one or more contributor license agreements.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file exposed on Github (readium) in the project repository.
+// ==LICENSE-END==
+
+const fromEntries = (it: Array<[string, any]>): any =>
+    it.reduce<any>(
+        (pv, [_key, _value]) =>
+            pv[_key] = typeof _value === "object" ? fromEntries(_value) : _value,
+        {});
+
+const ObjFromEntries = Object.fromEntries || fromEntries;
+
+export const deepClone = (o: unknown): any => {
```

Comment:
This function does not seem to work as intended, the clone is not structured as per the original object (I think `Object.fromEntries() should be called recursively too):

Repro test:

```javascript
const deepClone = (o) => {

    const _clone = (_o) =>
        Object.entries(_o).map(
            ([_key, _value]) =>
                typeof _value === "object" ? [_key, _clone(_value)] : [_key, _value]);

    const entries = _clone(o);
    return Object.fromEntries(entries);
};

const obj = {
    zbool: true,
    zstring: "str1",
    znumber: -1,
    zarray: [true, "str3", 111, {foo: "bar", deep: ["another array"]}, [1, 2, 3, {n:4}]],
    zobj: {
        zbool: false,
        zstring: "str2",
        znumber: 999,
        zarray: [false, "str5", 333, {num: 555, nest: [{nul: "null", und: undefined}]}, [1, 2, 3, {n:4}]],
    },
};
console.log(JSON.stringify(obj, null, 4));
console.log(JSON.stringify(deepClone(obj), null, 4));
```

Output:

```json
{
    "zbool": true,
    "zstring": "str1",
    "znumber": -1,
    "zarray": [
        true,
        "str3",
        111,
        {
            "foo": "bar",
            "deep": [
                "another array"
            ]
        },
        [
            1,
            2,
            3,
            {
                "n": 4
            }
        ]
    ],
    "zobj": {
        "zbool": false,
        "zstring": "str2",
        "znumber": 999,
        "zarray": [
            false,
            "str5",
            333,
            {
                "num": 555,
                "nest": [
                    {
                        "nul": "null"
                    }
                ]
            },
            [
                1,
                2,
                3,
                {
                    "n": 4
                }
            ]
        ]
    }
}
```json
```
{
    "zbool": true,
    "zstring": "str1",
    "znumber": -1,
    "zarray": [
        [
            "0",
            true
        ],
        [
            "1",
            "str3"
        ],
        [
            "2",
            111
        ],
        [
            "3",
            [
                [
                    "foo",
                    "bar"
                ],
                [
                    "deep",
                    [
                        [
                            "0",
                            "another array"
                        ]
                    ]
                ]
            ]
        ],
        [
            "4",
            [
                [
                    "0",
                    1
                ],
                [
                    "1",
                    2
                ],
                [
                    "2",
                    3
                ],
                [
                    "3",
                    [
                        [
                            "n",
                            4
                        ]
                    ]
                ]
            ]
        ]
    ],
    "zobj": [
        [
            "zbool",
            false
        ],
        [
            "zstring",
            "str2"
        ],
        [
            "znumber",
            999
        ],
        [
            "zarray",
            [
                [
                    "0",
                    false
                ],
                [
                    "1",
                    "str5"
                ],
                [
                    "2",
                    333
                ],
                [
                    "3",
                    [
                        [
                            "num",
                            555
                        ],
                        [
                            "nest",
                            [
                                [
                                    "0",
                                    [
                                        [
                                            "nul",
                                            "null"
                                        ],
                                        [
                                            "und",
                                            null
                                        ]
                                    ]
                                ]
                            ]
                        ]
                    ]
                ],
                [
                    "4",
                    [
                        [
                            "0",
                            1
                        ],
                        [
                            "1",
                            2
                        ],
                        [
                            "2",
                            3
                        ],
                        [
                            "3",
                            [
                                [
                                    "n",
                                    4
                                ]
                            ]
                        ]
                    ]
                ]
            ]
        ]
    ]
}
```

Reply:
thank you